src/Nethermind/Nethermind.Evm/Witness/VerkleExecWitness.cs

```diff
@@ -234,48 +231,69 @@ public byte[][] GetAccessedKeys()
-        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeSize, isWrite);
+        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeSize, ref gasAvailable, isWrite);
     }
 
-    private long AccessCodeHash(Address address, bool isWrite = false)
+    private bool AccessCodeHash(Address address, ref long gasAvailable, bool isWrite = false)
     {
-        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeHash, isWrite);
+        return AccessAccountSubTree(address, UInt256.Zero, AccountHeader.CodeHash, ref gasAvailable, isWrite);
     }
 
-    private long AccessAccountSubTree(Address address, UInt256 treeIndex, byte subIndex, bool isWrite = false)
+    private bool AccessAccountSubTree(Address address, UInt256 treeIndex, byte subIndex, ref long gasAvailable, bool isWrite = false)
     {
-        if (address.IsPrecompile(Osaka.Instance)) return 0;
-        return AccessKey(AccountHeader.GetTreeKey(address.Bytes, treeIndex, subIndex), isWrite);
+        if (address.IsPrecompile(Osaka.Instance)) return true;
+        return AccessKey(AccountHeader.GetTreeKey(address.Bytes, treeIndex, subIndex), ref gasAvailable, isWrite);
     }
 
-    private long AccessKey(Hash256 key, bool isWrite = false, bool leafExist = false)
+    private bool AccessKey(Hash256 key, ref long gasAvailable, bool isWrite = false, bool leafExist = false)
     {
-        long accessCost = 0;
-        if (_accessedLeaves.Add(key)) accessCost += GasCostOf.WitnessChunkRead;
-        if (_accessedSubtrees.Add(key.Bytes[..31].ToArray())) accessCost += GasCostOf.WitnessBranchRead;;
+        if (!_accessedLeaves.Contains(key))
+        {
+            if (gasAvailable < GasCostOf.WitnessChunkRead) return false;
+            _accessedLeaves.Add(key);
+            gasAvailable -= GasCostOf.WitnessChunkRead;
+        }
+
+        if (!_accessedSubtrees.Contains(key.Bytes[..31].ToArray()))
+        {
+            if (gasAvailable < GasCostOf.WitnessBranchRead) return false;
+            _accessedSubtrees.Add(key.Bytes[..31].ToArray());
+            gasAvailable -= GasCostOf.WitnessBranchRead;
+        }
 
-        if (isWrite)
+        if (!isWrite) return true;
+
+        if (!_modifiedLeaves.Contains(key))
+        {
+            if (gasAvailable < GasCostOf.WitnessChunkWrite) return false;
+            _modifiedLeaves.Add(key);
+            gasAvailable -= GasCostOf.WitnessChunkWrite;
+        }
+
+        if (!_modifiedSubtrees.Contains(key.Bytes[..31].ToArray()))
         {
-            if (_modifiedLeaves.Add(key)) accessCost += GasCostOf.WitnessChunkWrite;
-            if (_modifiedSubtrees.Add(key.Bytes[..31].ToArray())) accessCost += GasCostOf.WitnessBranchWrite;
+            if (gasAvailable < GasCostOf.WitnessBranchWrite) return false;
+            _modifiedSubtrees.Add(key.Bytes[..31].ToArray());
+            gasAvailable -= GasCostOf.WitnessBranchWrite;
         }
-        return accessCost;
+
+        return true;
```

Comment:
i think this should be a atomic operation - like only add anything to the witness when we have the gas to charge for both subtree and leaf

you can do the read and the write part seperatly I guess, but the leaf and subtree should be charged together

Reply:
changed to atomic operation